daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/layers/LayerFactory.scala

```diff
@@ -0,0 +1,326 @@
+   * and passing the parameter variables faster at runtime. This also does all the
+   * checking that the constructor has an argument for each variable of properly matching type, and
+   * has a getter for each result variable, again returning the proper type.
+   *
+   * This is called at schema compile time to ensure the layer code is properly defined and matches
+   * the layer variable definitions.
+   *
+   * It is called again at runtime after the Layer class is loaded by the SPI
+   * to ensure that the loaded layer class constructor signature at least matches the layer
+   * variables defined in the schema.
+   * @param lrd
+   * @param protoLayer the layer instance allocated by the SPI loader (zero-arg constructed)
+   * @return
+   */
+  def computeLayerVarsRuntime(lrd: LayerRuntimeData, protoLayer: Layer): LayerVarsRuntime = {
+    val optLayerVarsRuntime = alreadyCheckedLayers.get(protoLayer.name())
+    optLayerVarsRuntime.getOrElse {
+      // we know there is a default zero arg constructor
+      // find another constructor and check that there is an argument
+      // corresponding to each of the layer variables.
+      val c = protoLayer.getClass
+      val allConstructors = c.getConstructors.toSeq
+
+      val constructor: Constructor[_] =
+        if (allConstructors.length == 1) {
+          // There is only the default constructor.
+          // That's ok if there are no variables for the layer, which we check later.
+          allConstructors.head
+        } else if (allConstructors.length == 2) {
+          allConstructors.filter(_.getParameterCount > 0).head
+        } else {
+          def tooManyConstructorsMsg: String = {
+            s"""Layer class $c has multiple non-default constructors. It should have a default (no args)
+               | constructor and a single additional constructor with arguments for
+               | each of the layer's parameter variables.""".stripMargin
+          }
+          lrd.context.SDE(tooManyConstructorsMsg)
+        }
+
+      if (lrd.vmap.isEmpty && allConstructors.length == 1) {
+        // there are no vars, we're done
+        new LayerVarsRuntime(constructor, Nil, Nil)
+      } else {
+        // there is a constructor with args that are supposed to correspond to bound vars
+        val params = constructor.getParameters.toSeq
+        val nParams = params.length
+        val nVars = lrd.vmap.size
+
+        val paramTypes = constructor.getParameterTypes.toSeq
+        val paramVRDs = params.map { p =>
+          lrd.vmap.getOrElse(
+            p.getName,
+            lrd.context.SDE(s"No layer DFDL variable named '$p.getName' was found."),
+          )
+        }.toSeq
+
+        // Now we deal with the result getters and the corresponding vars
+        //
+        val allLayerVRDs = ListSet(lrd.vmap.toSeq.map(_._2): _*)
+        val returnVRDs = allLayerVRDs -- paramVRDs // set subtraction
```

Comment:
A thought, is there any value in saying that input variables are defined by by the constructor parameters and output variables are defined by the getters, and additional variables are ignored. This would allow layers to have variables that are sort of "helper" variables? For example, what about something like this:

**myLayer.dfdl.xsd**
```xml
...
<defineVariable name="foo" type="xs:int" defaultValue="2" />
<defineVariable name="bar" type="xs:int" defaultValue="{ $foo * 100}" />
...
```

The constructor might only accept and use `bar`, but a user to do something like this:
`daffodil parse -Dlyr:foo=10 ...`

So they could change the default value of `foo`, which would change the value of `bar`, which is what the layer actually uses. Or the user could just set `lyr:bar` directory if they want that extra control.

I'm not sure if that adds value or not. But it might simplify the implementation of this function since some of the logic here is making sure all defined variables are used?

I guess the argument against this is that a user might forget to use a variable and we want to alert them of that?